lint/variables.go

```diff
@@ -0,0 +1,176 @@
+			switch format {
+			case "seconds":
+				return strconv.FormatInt(val.Unix(), 10), nil
+			case "iso":
+				return val.Format(time.RFC3339), nil
+			default:
+				return "", fmt.Errorf("Unsupported momentjs time format: " + format)
+			}
+		default:
+			switch format {
+			case "date":
+				return val.Format(time.RFC3339), nil
+			default:
+				return strconv.FormatInt(val.UnixMilli(), 10), nil
+			}
+		}
+	default:
+		svalue := fmt.Sprintf("%s", value)
+		svalueList := []string{svalue, svalue, svalue}
+		// Implements https://grafana.com/docs/grafana/latest/variables/advanced-variable-format-options/
+		switch format {
+		case "csv":
+			return strings.Join(svalueList, ","), nil
+		case "doublequote":
+			return "\"" + strings.Join(svalueList, "\",\"") + "\"", nil
+		case "glob":
+			return "{" + strings.Join(svalueList, ",") + "}", nil
+		case "json":
+			data, err := json.Marshal(svalueList)
+			if err != nil {
+				return "", err
+			}
+			return string(data), nil
+		case "lucene":
+			return "(\"" + strings.Join(svalueList, "\" OR \"") + "\")", nil
+		case "percentencode":
+			return url.QueryEscape(strings.Join(svalueList, ",")), nil
+		case "pipe":
+			return strings.Join(svalueList, "|"), nil
+		case "raw":
+			return strings.Join(svalueList, ","), nil
+		case "regex":
+			return strings.Join(svalueList, "|"), nil
+		case "singlequote":
+			return "'" + strings.Join(svalueList, "','") + "'", nil
+		case "sqlstring":
+			return "'" + strings.Join(svalueList, "','") + "'", nil
+		case "text":
+			return strings.Join(svalueList, " + "), nil
+		case "queryparam":
+			values := url.Values{}
+			for _, svalue := range svalueList {
+				values.Add("var-"+name, svalue)
+			}
+			return values.Encode(), nil
+		default:
+			return svalue, nil
+		}
+	}
+}
+
+func variableSampleValue(s string) (string, error) {
+	var name, kind, format string
+	parts := strings.Split(s, ":")
+	switch len(parts) {
+	case 1:
+		// No format
+		name = s
+	case 2:
+		// Could be __from:date, variable:csv, ...
+		name = parts[0]
+		format = parts[1]
+	case 3:
+		// Could be __from:date:iso, ...
+		name = parts[0]
+		kind = parts[1]
+		format = parts[2]
+	}
+	// If it is part of the globals, return a string representation of a sample value
+	if value, ok := globalVariables[name]; ok {
+		return stringValue(name, value, kind, format)
+	}
+	// Assume variable type is a string
+	return stringValue(name, name, kind, format)
+}
+
+var variableMatchers = []struct {
+	pattern *regexp.Regexp
+	repl    func(string) (string, error)
+}{
+	{
+		regexp.MustCompile(`\$[[:word:]]+`),
+		func(s string) (string, error) {
+			return variableSampleValue(strings.TrimLeft(s, "$"))
+		},
+	},
+	{
+		regexp.MustCompile(`\$\{[^}]+\}`),
+		func(s string) (string, error) {
+			return variableSampleValue(strings.Trim(s, "${}"))
+		},
+	},
+	{
+		regexp.MustCompile(`\[\[[^\[\]]+\]\]`),
+		func(s string) (string, error) {
+			return variableSampleValue(strings.Trim(s, "[]"))
+		},
+	},
+}
```

Comment:
Given the body of the functions is so similar here, perhaps we could rephrase this?  I think you can get away without the string.Trims by using capturing groups in the regexs?  eg `\$(?<var>[[:word:]]+)` see https://github.com/google/re2/wiki/Syntax

You'll need to do something like https://elliotchance.medium.com/go-replace-string-with-regular-expression-callback-f89948bad0bb in expandVariables.

Reply:
I am honestly not fully convinced it will make it easier to maintain, but gave it a try. You tell me which one you prefer.